### PR TITLE
Raise better error if server doesn't return a full answer

### LINKED ISSUE #511: IndexError: bytearray index out of range
This happens for example using JavaServer.lookup() with the following ip: 135.148.100.83:25572

Traceback:
```
Traceback (most recent call last):
  File "/root/ServerSeeker/scan.py", line 42, in scan_chunk
    status = JavaServer.lookup(f"{ip}:{port}").status()
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/mcstatus/server.py", line 128, in status
    return self._retry_status(connection, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/mcstatus/utils.py", line 66, in sync_wrapper
    raise last_exc  # type: ignore # (This won't actually be unbound)
    ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/mcstatus/utils.py", line 62, in sync_wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/mcstatus/server.py", line 134, in _retry_status
    result = pinger.read_status()
             ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/mcstatus/pinger.py", line 116, in read_status
    if response.read_varint() != 0:
       ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/dist-packages/mcstatus/protocol/connection.py", line 254, in read_varint
    part = self.read(1)[0]
           ~~~~~~~~~~~~^^^
IndexError: bytearray index out of range
```

### PR DESCRIPTION
**Pull request by PerchunPak**
Closes #511.

### CODE REVIEW

File: mcstatus/protocol/connection.py [hunk 1]
```diff
@@ -448,6 +448,9 @@ def __init__(self) -> None:
 
     def read(self, length: int) -> bytearray:
         """Return :attr:`.received` up to length bytes, then cut received up to that point."""
+        if len(self.received) == 0:
+            raise IndexError("No data to read!")
```
Comment (ItsDrike) [inline mcstatus/protocol/connection.py:452]: While `IndexError` is indeed what would be raised if this happened without handling, do we really want to use this here? Wouldn't something like say an `IOError` make more sense? Indicating that there was an error in getting the output from the server.

File: mcstatus/protocol/connection.py [hunk 2]
```diff
@@ -448,6 +448,9 @@ def __init__(self) -> None:
 
     def read(self, length: int) -> bytearray:
         """Return :attr:`.received` up to length bytes, then cut received up to that point."""
+        if len(self.received) == 0:
```
Comment (ItsDrike) [inline mcstatus/protocol/connection.py:451]: Is `== 0` the right check to make here? Wouldn't `< length` make more sense?

File: mcstatus/protocol/connection.py (hunk 1 above)
Comment (PerchunPak) [reply mcstatus/protocol/connection.py:452]: This is what raising other classes, I though Liskov substitution principle should be applied here too.

Comment (ItsDrike) [reply mcstatus/protocol/connection.py]: Hm? What other classes? For example in `TCPSocketConnection`, the `read` method is already using `IOError`, somewhere, same thing for read_varint when the data isn't complete.

File: mcstatus/protocol/connection.py (hunk 2 above)
Comment (PerchunPak) [reply mcstatus/protocol/connection.py]: Fixed [here](https://github.com/py-mine/mcstatus/compare/7a22f5040d76f759ebb580ce719740da4ed0cf74..9c4af29f244827dc306cf1f9277f8b27d896abb4), thanks!

File: mcstatus/protocol/connection.py (hunk 1 above)
Comment (PerchunPak) [reply mcstatus/protocol/connection.py]: Changed here [here](https://github.com/py-mine/mcstatus/compare/77dfd352e41f95670578b9333ca511926d599420..9a6f4ad653e5ccc77ad0937407aa5f6f68bf35a1).

**Review by kevinkjt2000 — APPROVED** [review]

**Review by ItsDrike — APPROVED** [review]

**Pull request merged**